include/deal.II/grid/tria_accessor.templates.h

```diff
@@ -2040,6 +2040,159 @@ TriaAccessor<structdim, dim, spacedim>::diameter () const
+  if (structdim==1)
+    return std::make_pair( (this->vertex(1)+this->vertex(0))*0.5,
+                           (this->vertex(1)-this->vertex(0)).norm()*0.5);
+
+  // Each element in vertices_within_ball is the bool value
+  // corresponding to whether the vertex is inside the ball.
+  std::array<bool, GeometryInfo<structdim>::vertices_per_cell> vertices_within_ball;
+
+  //First let all the vertices be outside
+  std::fill(vertices_within_ball.begin(),
+            vertices_within_ball.end(),
+            false);
+
+  // Get an initial guess by looking at the largest diagonal
+  Point<spacedim> center;
+  double radius = 0;
+
+  switch (structdim)
+    {
+    case 2:
+    {
+      const Point<spacedim> p30( this->vertex(3)-this->vertex(0));
+      const Point<spacedim> p21( this->vertex(2)-this->vertex(1));
+      if (p30.norm() > p21.norm())
+        {
+          center = this->vertex(0) + 0.5* p30;
+          radius = p30.norm()/2.;
+          vertices_within_ball[3] = true;
+          vertices_within_ball[0] = true;
+        }
+      else
+        {
+          center = this->vertex(1) + 0.5* p21;
+          radius = p21.norm()/2.;
+          vertices_within_ball[2] = true;
+          vertices_within_ball[1] = true;
+        }
+      break;
+    }
+    case 3:
+    {
+      const Point<spacedim> p70( this->vertex(7)-this->vertex(0));
+      const Point<spacedim> p61( this->vertex(6)-this->vertex(1));
+      const Point<spacedim> p25( this->vertex(2)-this->vertex(5));
+      const Point<spacedim> p34( this->vertex(3)-this->vertex(4));
+      const std::vector<double> diagonals= { p70.norm(),
+                                             p61.norm(),
+                                             p25.norm(),
+                                             p34.norm()
+                                           };
+      const std::vector<double>::const_iterator
+      it = std::max_element( diagonals.begin(), diagonals.end());
+      if (it == diagonals.begin())
+        {
+          center = this->vertex(0) + 0.5* p70;
+          vertices_within_ball[7] = true;
+          vertices_within_ball[0] = true;
+        }
+      else if (it == diagonals.begin()+1)
+        {
+          center = this->vertex(1) + 0.5* p61;
+          vertices_within_ball[6] = true;
+          vertices_within_ball[1] = true;
+        }
+      else if (it == diagonals.begin()+2)
+        {
+          center = this->vertex(5) + 0.5* p25;
+          vertices_within_ball[2] = true;
+          vertices_within_ball[5] = true;
+        }
+      else
+        {
+          center = this->vertex(4) + 0.5* p34;
+          vertices_within_ball[3] = true;
+          vertices_within_ball[4] = true;
+        }
+      radius = *it * 0.5;
+      break;
+    }
+    default:
+      Assert (false, ExcNotImplemented());
+      return std::pair<Point<spacedim>,double>();
+    }
+
+  // For each vertex that is found to be geometrically outside the ball
+  // enlarge the ball.
```

Comment:
-> enlarge the ball so that the new ball contains both the previous ball and the given vertex